File: workflow-scheduler/app/app/schemas/job.py
from typing import Optional, Tuple, List
from datetime import datetime
from enum import IntEnum

from pydantic import BaseModel


# Shared properties
class JobBase(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    owner: Optional[str] = None
    owner_id: Optional[int] = None

    # workflow: Optional[Json] = None

    # workflow: WorkflowInDB = None
    workflows: Optional[List[Tuple[int, str, datetime]]] = []   # [workflow_id, workflow_type, workflow_execute_at]
    list_protocol_and_database: Optional[List[Tuple[int, int]]] = []   # [(protocol_id, database_id)]
    dataflow_path: Optional[str]

    # data_acquisition_protocol: Optional[Json] = None
    # data_acquisition_protocol_execute_at: Optional[datetime]
    # dataflow: Optional[Json] = None
    # dataflow_type: Optional[str] = None
    # dataflow_execute_at: Optional[datetime]
    # database: Optional[UUID]
    # service_bookings:
    execute_at: Optional[datetime] = None
    created_at: Optional[datetime] = None


# Properties to receive on item creation
class JobCreate(JobBase):
    title: str
    owner: str
    owner_id: int
    # created_at: datetime  # = datetime.now()
    # uuid: UUID = uuid4()
    # workflow: WorkflowInDB
    created_at: datetime


# Properties to receive on item update
class JobUpdate(JobBase):
    pass


# Properties shared by models stored in DB
class JobInDBBase(JobBase):
    id: int
    title: str
    owner: str
    owner_id: int
    created_at: datetime

    class Config:
        orm_mode = True


# Properties to return to client
class Job(JobInDBBase):
    pass


# Properties properties stored in DB
class JobInDB(JobInDBBase):
    pass

File: frontend/src/app/_models/job.ts
import {
    Workflow, WorkflowInfoList, WorkflowInfoTuple,
    Dataflow, DataflowInfoList,
    DataProtocol, DataProtocolInfoList,
    Protocol, Database,
} from '@app/_models';


export interface Job {
    id?: number;
    // uuid?: string;
    title: string;
    description?: string;
    owner?: string;
    owner_id?: number;

    workflows?: Workflow[];
    dataflows?: Dataflow[];
    // data_protocols?: DataProtocol[];
    list_protocol_and_database?: ProtocolAndDatabaseTuple[];

    execute_at: Date;
    created_at: Date;
    running: boolean;
    database: string;
}

export interface JobInfo {
    id?: number;
    // uuid?: string;
    title: string;
    description?: string;
    owner?: string;
    owner_id?: number;

    workflows?: WorkflowInfoTuple[];  // WorkflowInfoList[];
    dataflows?: DataflowInfoList;
    // data_protocols?: DataProtocolInfoList;
    list_protocol_and_database?: [number, number][];

    execute_at: Date;
    created_at: Date;
    running: boolean;
    database?: string;
}

export interface JobInfoList {
    data: JobInfo[];
}

export interface JobStatus {
    online: boolean;
    status: string;
}

export interface ProtocolAndDatabaseTuple {
    protocol: Protocol;
    database: Database;
}
